### PR TITLE
(HDP-161) Remove `submit_only` server URL when it times out repeatedly (6.x)

### PR DESCRIPTION
**Pull request by Magisus**
This commit adds logic to PDB's Ruby http utils to track timeouts for
the `submit_only` server URLs, which are used by HDP. If a given URL
times out twice in a row, it is removed from the loaded config, to avoid
tying up JRubies indefinitely in the case where HDP has gone down. By
removing the URL, PDB will no longer try to submit things to it,
allowing requests to the other configured URLs to proceed quickly.

The URL will be re-added to the config whenever the JRuby instance is
refreshed: with a restart of Puppet Server, when the JRuby instance is
recreated when the "max requests per instance" threshold is hit, or if
the pool is flushed manually. Otherwise, the misbehaving URL will be
kept out of the list permanently.

### CODE REVIEW

**Comment by Magisus** [comment]
Note: each Jruby only processes on request at a time, so there are no race conditions in this code. However each JRuby will will track timeouts and remove the URL from its loaded config independently.

**Comment by Magisus** [comment]
The Docker tests appear to be failing all the time right now, unrelated to this change.

File: puppet/lib/puppet/util/puppetdb/http.rb [hunk 1]
```diff
@@ -141,7 +168,14 @@ def self.failover_action(path_suffix, server_urls, sticky, ssl_context = nil, ht
           end
 
           response = Timeout.timeout(config.server_url_timeout) do
-            http_callback.call(http, route)
+            response = http_callback.call(http, route)
+
+            if config.submit_only_server_urls.include?(server_url)
+              # Clear any previously registered timeout for this URL
+              @@submit_only_url_timeouts[server_url] = false
```
Comment (austb) [inline puppet/lib/puppet/util/puppetdb/http.rb:175]: I believe we will need to copy this logic in `broadcast_action` as well. It is configurable in the `action` method below which one a command will use.

Comment (Magisus) [reply puppet/lib/puppet/util/puppetdb/http.rb:175]: Okay, yeah found a testing hole now that you point this out, also. Working to fix that up.

**Comment by Magisus** [comment]
Okay I made sure to add tests for both `broadcast_command=true/false`. The ones for `false` are a little funky, since it apparently doesn't try the `submit_only` url unless the ones before it both fail, which means that all URLs need to fail to get to testing the `submit_only` timeout.

**Pull request merged**